src/osp/hierarchical_bitset.h

```diff
@@ -0,0 +1,607 @@
+{
+    static_assert(std::is_unsigned_v<INT_T>);
+    return (block & (INT_T(1) << bit) ) != 0;
+}
+
+/**
+ * @brief Copy a certain number of bits from one int array to another
+ *
+ * This function treats integer arrays as bit arrays. Bits are indexed from
+ * LSB to MSB.
+ *
+ * TODO: maybe add a starting bit offset too?
+ *
+ * @param pSrc   [in] Integer array to copy
+ * @param pDest  [out] Integer array to write into
+ * @param bits   [in] Number of bits to write
+ */
+template<typename INT_T>
+constexpr void copy_bits(INT_T const* pSrc, INT_T* pDest, size_t bits) noexcept
+{
+    static_assert(std::is_unsigned_v<INT_T>);
+
+    while (bits >= sizeof(INT_T) * 8)
+    {
+        *pDest = *pSrc;
+        pSrc ++;
+        pDest ++;
+        bits -= sizeof(INT_T) * 8;
+    }
+
+    if (bits != 0)
+    {
+        INT_T const mask = (~INT_T(0)) << bits;
+
+        *pDest &= mask;
+        *pDest |= ( (~mask) & (*pSrc) );
+    }
+}
+
+template<typename INT_T>
+constexpr void set_bits(INT_T* pDest, size_t bits) noexcept
+{
+    static_assert(std::is_unsigned_v<INT_T>);
+
+    constexpr INT_T c_allOnes = ~INT_T(0);
+
+    while (bits >= sizeof(INT_T) * 8)
+    {
+        *pDest = c_allOnes;
+        pDest ++;
+        bits -= sizeof(INT_T) * 8;
+    }
+
+    if (bits != 0)
+    {
+        *pDest |= ~( c_allOnes << bits );
+    }
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief A bitset featuring hierarchical rows for fast iteration over set bits
+ *
+ * This container requires rows of bits, divided into blocks. This is
+ * represented as an array of unsigned integer types, where each int is a block.
+ *
+ * The main advantage of using an int, is that their bits can be tested for
+ * ones all at once, by comparing the int to zero.
+ *
+ * There are multiple rows. The bottom row [0] is the user's bits that can be
+ * set and reset. For the rest of the rows above, each bit N is set if the row
+ * below's block N is non-zero.
+ *
+ * The bottom row sets the size of the container, and each row above must be
+ * be large enough to have a bit for each block of the row below.
+ *
+ * For example, when using 8-bit block sizes, and 40 bits for container size:
+ *
+ * Row 0 size: 40 bits  (5 blocks needed)
+ * Row 1 size: 5 bits   (1 block)
+ * Row 2 size: 0, not needed
+ * ...
+ *
+ * To represent a set of integers: { 0, 1, 18, 19 }:
+ *
+ * Row 1: 00000101
+ * Row 0: 00000011 00000000 00001100 00000000 00000000
+ *
+ * * Row 1's first 5 bits corresponds to Row 0's blocks.
+ * * Row 1's bit 0 and bit 2 means Row 0's block 0 and block 2 is non-zero
+ *
+ * note: Bits are displayed in msb->lsb, but arrays are left->right
+ *
+ * Iterating or searching for set bits on the bottom row is as simple as looking
+ * up the set bits in the blocks the top row, and recursing down.
+ *
+ * This operation is O(number of rows), or O(log n)
+ */
+template <typename BLOCK_INT_T = uint64_t>
+class HierarchicalBitset
```

Comment:
Hard for me to comment much on the logic just by inspection, but this is definitely a very cool data structure.